tests/integration/app-charm/src/charm.py

```diff
@@ -242,6 +275,51 @@ def _http_request(self, event) -> None:
         response = http_request(credentials, endpoint, method, payload)
         event.set_results({"results": json.dumps(response)})
 
+    def _generate_cert(self, event) -> None:
+        """Handle the action that generates a certificate."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit.")
+            return
+
+        # read parameters from the event
+        common_name: str = event.params["common-name"]
+
+        # generate the certificate
+        cert, key = generate_cert(common_name)
+        # save the certificate and key to disk
+        etcd_snap_dir = Path(ETCD_SNAP_DIR)
+        etcd_snap_dir.mkdir(parents=True, exist_ok=True)
+        Path(etcd_snap_dir / "client.pem").write_text(cert)
+        Path(etcd_snap_dir / "client.key").write_text(key)
+        # set the results of the action
+        event.set_results({"certificate": cert, "key": key})
+
+    @retry(stop=stop_after_attempt(3), wait=wait_fixed(5), reraise=True)
+    def _install_etcd_snap(self) -> bool:
+        """Install the etcd snap."""
+        try:
+            self.etcd_snap.ensure(snap.SnapState.Present, channel="3.5/edge")
+            self.etcd_snap.hold()
+            return True
+        except snap.SnapError as e:
+            logger.error(str(e))
+            return False
```

Comment:
This will not work as intended, if the goal is to return a boolean, you should:
- remove `reraise=True`
- add the tenacity retry inline - not as decorator

```suggestion
    def _install_etcd_snap(self) -> bool:
        """Install the etcd snap."""
        for attempt in Retrying(stop=stop_after_attempt(3), wait=wait_fixed(5)):
            with attempt:
                self.etcd_snap.ensure(snap.SnapState.Present, channel="3.5/edge")
                self.etcd_snap.hold()
                return True
                
        logger.error("Failed to install etcd snap after retries.")
        return False
```

if needed to log the error, add an try/except Exception where you log on the except and raise